### PR TITLE
Fix NaN in Gemma3/EmbeddingGemma when batching mixed-length sequences…

### LINKED ISSUE #45491: [Gemma3] NaN embeddings on GPU when batching sequences of mixed length (sliding window attention + all-padding windows)
### System Info

**System Info**

- `transformers`: 4.45.1
- `sentence-transformers`: 5.1.2
- `tokenizers`: 0.20.0
- `safetensors`: 0.4.5
- `PyTorch`: ≥ 2.6.0
- Serving runtime: `pytorch/torchserve-kfs:0.12.0` (Python 3.9, Linux x86_64)
- GPU: NVIDIA (CUDA, via KServe / TorchServe on Kubernetes)
- CPU inference: **not affected**

---

**Description**

When encoding a batch of sentences of **mixed lengths** with `google/EmbeddingGemma` (300M) via `SentenceTransformer.encode_query()`, short sentences receive an **all-NaN embedding vector** on GPU. The same inputs encoded individually (`batch_size=1`) return correct, fully-normalized embeddings.

**Trigger condition:** the shortest sequence in the batch must be short enough that, when padded to the length of the longest sequence, one or more sliding-window attention windows fall **entirely on padding tokens**.

In the reported case:
- Chunk 0: **142 tokens** → padded to 728 tokens → **NaN=768/768** in the batch
- Chunks 1–3: 452 / 555 / 728 tokens → valid embeddings

The model's sliding window size is **512**. With 728 − 142 = 586 padding tokens appended, the first attention windows of Chunk 0 are 100% padding, causing `softmax([-inf, -inf, …]) = 0/0 = NaN` in the GPU kernel — a silent numerical error (no exception raised, no warning).

**Actual behavior:** Chunk 0 (shortest, ~142 tokens) returns a vector of 768 NaN values when batched with longer sequences. No exception or warning is raised.

**Workaround:** detect NaN after encoding and re-encode the affected inputs with `batch_size=1`.

### Who can help?

@ArthurZucker 

@Cyrilvallez 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

**Steps to reproduce**

```python
from sentence_transformers import SentenceTransformer
import numpy as np

model = SentenceTransformer("google/EmbeddingGemma-300M")  # GPU required

chunks = [
    "ipso literam...",  # ~142 tokens
    "ipso literam...",                      # ~452 tokens
    "ipso literam...",           # ~555 tokens
    "ipso literam...",                # ~728 tokens
]

# Batch — Chunk 0 produces all-NaN on GPU
embs = model.encode_query(chunks, batch_size=64)
for i, emb in enumerate(embs):
    print(f"Chunk {i}: NaN={np.any(np.isnan(emb))}")
# Output:
# Chunk 0: NaN=True   ← 768/768 NaN values
# Chunk 1: NaN=False
# Chunk 2: NaN=False
# Chunk 3: NaN=False

# Individual — all correct
for chunk in chunks:
    emb = model.encode_query([chunk], batch_size=1)
    print(f"NaN={np.any(np.isnan(emb))}")  # False for all
```

### Expected behavior

**Expected behavior:** all embeddings in the batch are valid (no NaN).

### PR DESCRIPTION
**Pull request by Kabir08**
This fixes the NaN issue when batching mixed-length sequences with sliding window attention (e.g., with `EmbeddingGemma` / Gemma3 models). 

The patch ensures that when a query position's entire sliding window falls within the padding region (all keys masked with `-inf`), the attention bias/mask is adjusted to allow all keys (bias=0). This produces a valid uniform softmax distribution instead of `softmax([-inf, ...]) = NaN`. 

This matches the intended behavior for padding positions, which are always excluded from downstream pooling/embedding computation and thus do not affect the final output. The change prevents silent numerical instability on certain GPU SDPA backends and in eager-mode attention, while keeping CPU behavior and single-sequence (`batch_size=1`) results unchanged.

Fixes #45491.

### What does this PR do?

When a short sequence is padded to match a much longer one in a batch (common in batched inference with variable-length inputs), some query positions in the short sequence may have their entire sliding window (e.g., size 512 for Gemma3 local attention layers) fall completely within the padding tokens of the longer sequences.

This results in an attention score row of all `-inf`, causing `softmax` to produce `NaN` (0/0) on affected GPU kernels and in some eager implementations. The NaNs then propagate to the final embeddings (e.g., all-NaN vectors for the shortest items in a batch when using `SentenceTransformer.encode_query()` or similar).

The fix detects these all-masked rows in the sliding window mask construction and sets the bias to 0 for those positions, yielding a uniform attention distribution. Since these positions correspond exclusively to padding tokens, they are masked out or ignored in subsequent mean-pooling / embedding extraction steps, so the output remains correct and identical to per-example encoding.

This resolves the reported bug for `google/EmbeddingGemma-300M` (and other Gemma3 variants using sliding window attention) on GPU with mixed-length batches, without changing behavior for valid (non-padding) positions or non-sliding-window models.

### Related issues / context
- Root cause: Sliding window attention + dynamic padding in batched processing → all-padding windows → numerical instability in softmax.
- Affects: Models with `sliding_window` in their config (Gemma3 local layers, window size typically 512 or similar).
- Previously worked around by forcing `batch_size=1` or re-encoding NaN results individually.

No new dependencies. No breaking changes.

## Code Agent Policy
The Transformers repo is currently being overwhelmed by a large number of PRs and issue comments written by code agents. ...

- [x] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.  
  → Discussed in [#45491](https://github.com/huggingface/transformers/issues/45491)
- [ ] Did you make sure to update the documentation with your changes? ...
- [ ] Did you write any new necessary tests?

## Who can review?
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

**Recommended tags (attention + Gemma-related maintainers — keep under 3 where possible):**
- @ArthurZucker (text models, attention, Gemma expertise)
- @Cyrilvallez (text models, model loading, attention)
- @vasqu (attention)

### CODE REVIEW

**Comment by vasqu** [comment]
See my comment here: https://github.com/huggingface/transformers/issues/45491#issuecomment-4281075391

This is likely a padding side issue, not really a transformers issue

**Comment by Kabir08** [comment]
Understood, i'm closing the pr
